ckb-importer/src/main/java/com/hartwig/serve/ckb/datamodel/variant/VariantFactory.java

```diff
@@ -16,72 +18,74 @@
+    public VariantFactory(@NotNull CkbJsonDatabase ckbJsonDatabase) {
+        this.ckbJsonDatabase = ckbJsonDatabase;
+        variantsById = ckbJsonDatabase.variants().stream().collect(Collectors.toMap(JsonVariant::id, variant -> variant));
+        genesById = ckbJsonDatabase.genes().stream().collect(Collectors.toMap(JsonGene::id, gene -> gene));
     }
 
     @NotNull
-    public static List<Variant> extractVariants(@NotNull CkbJsonDatabase ckbJsonDatabase, @NotNull List<VariantInfo> variantInfos) {
-        List<Variant> variants = Lists.newArrayList();
-        for (VariantInfo variantInfo : variantInfos) {
-            variants.add(resolveVariant(ckbJsonDatabase, variantInfo));
-        }
-
-        return variants;
+    public List<Variant> extractVariants(@NotNull List<VariantInfo> variantInfos) {
+        return variantInfos.stream().map(this::resolveVariant).collect(Collectors.toList());
     }
 
     @NotNull
-    private static Variant resolveVariant(@NotNull CkbJsonDatabase ckbJsonDatabase, @NotNull VariantInfo variantInfo) {
-        for (JsonVariant variant : ckbJsonDatabase.variants()) {
-            if (variant.id() == variantInfo.id()) {
-                return ImmutableVariant.builder()
-                        .id(variant.id())
-                        .createDate(variant.createDate())
-                        .updateDate(variant.updateDate())
-                        .fullName(variant.fullName())
-                        .variant(variant.variant())
-                        .impact(variant.impact())
-                        .proteinEffect(variant.proteinEffect())
-                        .type(variant.type())
-                        .associatedWithDrugResistance(variant.associatedWithDrugResistance())
-                        .transformingActivity(variant.transformingActivity())
-                        .polymorphism(variant.polymorphism())
-                        .hotspotReference(variant.hotspotReference())
-                        .isHotspot(variant.hotspotReference() != null)
-                        .gene(resolveGene(ckbJsonDatabase, variant.gene()))
-                        .referenceTranscriptCoordinate(convertReferenceTranscriptCoordinate(variant.referenceTranscriptCoordinate()))
-                        .allTranscriptCoordinates(convertAllTranscriptCoordinates(variant.allTranscriptCoordinates()))
-                        .categoryVariantPaths(convertCategoryVariantPaths(variant.categoryVariantPaths()))
-                        .memberVariants(convertMemberVariants(variant.memberVariants()))
-                        .description(ReferenceFactory.extractDescription("variant", variant.id(), variant.descriptions()))
-                        .references(ReferenceFactory.extractDescriptionReferences(ckbJsonDatabase, variant.descriptions()))
-                        .build();
-            }
+    private Variant resolveVariant(@NotNull VariantInfo variantInfo) {
+        if (variantsById.containsKey(variantInfo.id())) {
```

Comment:
should this not equals? Otherwise id of e.g. 10 versus 100 can be also true

Reply:
`variantsById` is a `Map<Integer, JsonVariant>`, so [containsKey()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#containsKey-java.lang.Object-) will check for the presence of that specific integer in the map's keyset (unlike `String.contains()`). I introduced a map here because they allow very fast random access for known keys, which means each variant can be resolved almost instantaneously. This is much more efficient than the previous approach of requiring a full scan of all variants in ckbJsonDatabase for every `variantInfo` that needs to be resolved.